File: src/test/java/nz/co/searchwellington/controllers/SubmissionProcessingServiceTest.java
package nz.co.searchwellington.controllers;

import nz.co.searchwellington.geocoding.GeoCodeService;
import nz.co.searchwellington.model.Feed;
import nz.co.searchwellington.model.Newsitem;
import nz.co.searchwellington.model.UrlWordsGenerator;
import nz.co.searchwellington.model.User;
import nz.co.searchwellington.repositories.HandTaggingDAO;
import nz.co.searchwellington.repositories.ResourceRepository;
import nz.co.searchwellington.repositories.TagDAO;
import nz.co.searchwellington.utils.UrlCleaner;

import org.apache.struts.mock.MockHttpServletRequest;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.mockito.MockitoAnnotations;

public class SubmissionProcessingServiceTest {

	private static final String FEED_NAME = "A feed";
	@Mock UrlCleaner urlCleaner;
	@Mock GeoCodeService geocodeService;
	@Mock TagDAO tagDAO;
	@Mock HandTaggingDAO tagVoteDAO;
	@Mock ResourceRepository resourceDAO;
	@Mock Newsitem resource;
	
	private MockHttpServletRequest request;
	@Mock Feed feed;
	@Mock User loggedInUser;
	
	@Before
	public void setup() {
		MockitoAnnotations.initMocks(this);
		request = new MockHttpServletRequest();
		Mockito.when(feed.getName()).thenReturn(FEED_NAME);
	}
	
	@Test
	public void shouldFlattenLoudCapsHeadlinesInUserSubmissions() throws Exception {
		SubmissionProcessingService submissionProcessingService = new SubmissionProcessingService(urlCleaner, geocodeService, tagDAO, tagVoteDAO, resourceDAO);
		
		request.addParameter("title", "THE QUICK BROWN FOX JUMPED OVER THE LAZY DOG");
		submissionProcessingService.processTitle(request, resource);
		
		Mockito.verify(resource).setName("The quick brown fox jumped over the lazy dog");
	}
	
	@Test
	public void shouldPopulateAcceptanceFieldsOnInitalSubmission() throws Exception {
		SubmissionProcessingService submissionProcessingService = new SubmissionProcessingService(urlCleaner, geocodeService, tagDAO, tagVoteDAO, resourceDAO);
		request.addParameter("acceptedFromFeed", UrlWordsGenerator.makeUrlWordsFromName(FEED_NAME));
		
		Mockito.when(resourceDAO.loadFeedByUrlWords("a-feed")).thenReturn(feed);
		submissionProcessingService.processAcceptance(request, resource, loggedInUser);
		
		Mockito.verify(resource).setFeed(feed);
		Mockito.verify(resource).setAcceptedBy(loggedInUser);
		// TODO assert acceptance time was set!
	}
	
}

File: src/test/scala/nz/co/searchwellington/controllers/SubmissionProcessingServiceTest.scala
package nz.co.searchwellington.controllers

import nz.co.searchwellington.feeds.PlaceToGeocodeMapper
import nz.co.searchwellington.geocoding.osm.{GeoCodeService, OsmIdParser}
import nz.co.searchwellington.model.{Feed, Newsitem, UrlWordsGenerator, User}
import nz.co.searchwellington.repositories.mongo.MongoRepository
import nz.co.searchwellington.repositories.{HandTaggingDAO, TagDAO}
import nz.co.searchwellington.urls.UrlCleaner
import org.joda.time.DateTimeZone
import org.junit.Before
import org.mockito.{Mock, MockitoAnnotations}
import org.springframework.mock.web.MockHttpServletRequest
import uk.co.eelpieconsulting.common.dates.DateFormatter

import java.util.UUID

class SubmissionProcessingServiceTest {

  private val FEED_NAME: String = "A feed"

  @Mock val urlCleaner: UrlCleaner = null
  @Mock val geocodeService: GeoCodeService = null
  @Mock val tagDAO: TagDAO = null
  @Mock val mongoRepository: MongoRepository = null
  @Mock val tagVoteDAO: HandTaggingDAO = null
  @Mock val resource: Newsitem = null
  val feed = Feed(id = UUID.randomUUID().toString, title = Some(FEED_NAME))
  @Mock val loggedInUser: User = null
  @Mock val osmIdParser: OsmIdParser = null
  @Mock val placeToGeocodeMapper: PlaceToGeocodeMapper = null

  private var urlWordsGenerator: UrlWordsGenerator = null
  private var request: MockHttpServletRequest = null
  private var submissionProcessingService: SubmissionProcessingService = null

  @Before def setup {
    MockitoAnnotations.initMocks(this)
    request = new MockHttpServletRequest
    urlWordsGenerator = new UrlWordsGenerator(new DateFormatter(DateTimeZone.UTC))
    submissionProcessingService = new SubmissionProcessingService(tagDAO, tagVoteDAO, osmIdParser, placeToGeocodeMapper, mongoRepository)
  }

}
